Classic McEliece Cryptodoc (#186)
sehlen-bsi/botan-docs: docs/cryptodoc/src/05_10_cmce.rst

   +----------------------------+-----------+-----------+-----------+----------------------------+---------------------------------+
   | ``mceliece8192128<pc><f>`` | 13        | 8192      | 128       | :math:`z^{13}+z^4+z^3+z+1` | :math:`y^{128}+y^7+y^2+y+1`     |
   +----------------------------+-----------+-----------+-----------+----------------------------+---------------------------------+

.. _pubkey/cmce/key_generation:

Algorithm Internals
-------------------

The implementation of Classic McEliece is organized based on the algorithm's
operations and structures. Table :ref:`Classic McEliece components
<pubkey/cmce/component_table>` shows an overview of the components and
their corresponding file locations. Classic McEliece, like all Key Encapsulation
Mechanisms (KEMs), consists of three main algorithms: Key generation, key
encapsulation, and key decapsulation. In the case of Classic McEliece, the key
generation process involves creating a field ordering and an irreducible
polynomial. These two components define the randomly chosen Goppa code, the
secret information stored in the private key. A parity check matrix in

> systematic form is derived from the Goppa code to generate the public key used

Author: Well, the parity check matrix is always in systematic form in the end. Only the transformation to it uses the semi-systematic form. After `move_columns`, the matrix becomes systematic.

> systematic form is derived from the Goppa code to generate the public key used

Reviewer: True

   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+
   | :ref:`Field Ordering <pubkey/cmce/field_ordering>`     | :srcref:`[src/lib/pubkey/classic_mceliece]/cmce_field_ordering.h`| Field ordering creation and Beneš network computation                      |
   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+
   | :ref:`Polynomial Arithmetic <pubkey/cmce/poly>`        | :srcref:`[src/lib/pubkey/classic_mceliece]/cmce_poly.h`          | Polynomial data structure, logic, and minimal polynomial creation          |
   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+

>    | :ref:`Bit Vector Arithmetic <pubkey/cmce/bitvector>`   | :srcref:`src/lib/utils/bitvector.h`                              | Representation of bit-vectors and bit-vector arithmetic                    |

Author: Good point!

algorithms, and bit-vectors with different semantic contexts. Specific integers,
like raw Galois Field elements and their modulus, are
also represented as strong types. This reliance on strong types is similar to using them
in :ref:`SPHINCS+ <signatures/sphincsplus/types>` for
similar reasons. The Classic McEliece implementation enhances readability,
clarity, and reliability by employing strong types.

..  _pubkey/cmce/gf:

Galois Field Arithmetic
^^^^^^^^^^^^^^^^^^^^^^^

The Goppa code is based on the Galois field (GF) :math:`\mathbb{F}_{q}`, where :math:`q=2^m`.
To represent GF elements, we utilize the class ``Classic_McEliece_GF``. Each
element is defined by two polynomials: one defines its value, and the other
defines the modulus of the field. These values are represented as unsigned
integers. For example, the element :math:`z^3+z+1` is represented by the unsigned
integer ``0b1011``.

The GF elements support various operations such as addition, multiplication,
and inversion. These operations are implemented in constant time for operands
within the same field. Multiplication is performed using a constant-time

> schoolbook multiplication algorithm with a consecutive constant-time reduction.

Author: As far as I know _schoolbook multiplication_ is a well known term. However, the term _long multiplication_ seems to be more scientific. I'll change it 👍

> schoolbook multiplication algorithm with a consecutive constant-time reduction.

Reviewer: 👍

I guess it was not the "schoolbook" that bothered me but the "a [...] algorithm" which seemed ambiguous

Inversion is achieved using Fermat's little theorem: :math:`a^{-1} = a^{q-2}`.
For that, exponentiation is implemented using a simple square-and-multiply

> algorithm. It is worth noting that square-and-multiply has no timing

Reviewer: Then I'd just say "a constant-time squate-and-multiply", which is fair because in this context the algorithm is constant time as the exponent is constant. (I just expect comments on whether the algorithm is implemented constant-time if we mention just the algorithm).

side-channel vulnerabilities since the exponent remains constant for each field.

.. _pubkey/cmce/field_ordering:

Field Ordering
^^^^^^^^^^^^^^

Section 8.2 of [CMCE-ISO]_ defines the field ordering as a permutation
of elements in the Galois field :math:`\mathbb{F}_{q}`. This ordering is
required for generating a random sequence :math:`\alpha_0,...,\alpha_{n-1}` of distinct
:math:`\mathbb{F}_{q}` elements, which is necessary for creating the Goppa
code. Botan's ``Classic_McEliece_Field_Ordering`` serves as a container for the
field ordering and includes the algorithm to generate it. To sort the pairs
:math:`(a_i, i)` (as described in Step 3 of Section 8.2 in [CMCE-ISO]_), a
constant-time bitonic sort implementation is utilized. This sorting algorithm is

> particularly suitable for sets with a power of two elements.

Author: I think you could also use bitonic sort with non 2^n set by appending dummy elements (e.g. zeros) and removing them after sorting. I think the wording is good enough.

> particularly suitable for sets with a power of two elements.

Reviewer: Sure, I agree that this is too much detail here

Another vital role of the field ordering class is managing the Beneš network.
As Section 9.2.10 of [CMCE-ISO]_ outlines, the Beneš network stores the field
ordering as control bits in a compact form. Botan employs
a constant-time implementation of the ``controlbits`` algorithm presented in

> Fig. 7.1 of [CBits]_ to create control bits, and the ``permutation`` algorithm

Author: CMCE Implementation guide:
>See [9] for Python software and computer-verified correctness proofs for control-bit computation; the Python software is also included in the Classic McEliece specification.

They simply copied the paper python code for the ISO spec. 

> Also, should we mention that the class provides the alpha/pi values and that it therefore satisfies all algorithms of ISO Section 8.2?

Yeah, good idea!

> Fig. 7.1 of [CBits]_ to create control bits, and the ``permutation`` algorithm

Reviewer: Ah ok, I thought the sentence meant that the CBits reference only applies to the controlbits algorithm and not also the `permutation`/`create_from_control_bits` algorithms. I'd suggest sth like:

`Botan employs constant-time implementations of the controlbits and permutation algorithms presented in Fig. 7.1 of [CBits]_ to create control bits and to reconstruct the field ordering from given control bits, respectively.`

to reconstruct the field ordering from the control bits. The creation
algorithm also utilizes constant-time bitonic sorting.

.. _pubkey/cmce/poly:

Polynomial Arithmetic
^^^^^^^^^^^^^^^^^^^^^

The Classic McEliece algorithm uses elements in the polynomial ring

> :math:`\mathbb{F}_q [y]` in multiple places. The monic irreducible

Author: At this point I only want to say: Look, polynomials are used for multiple different cases. I admit the wording can be improved!

polynomial defines the Goppa code, and performing decapsulation requires an
error locator polynomial. Botan's ``Classic_McEliece_Poly`` class is designed
to represent these polynomials. It consists of a vector of GF elements, which
serve as the polynomial coefficients. Additionally, this class provides
the necessary logic to evaluate the polynomial at a given point in
:math:`\mathbb{F}_q`.

> Polynomial arithmetic is required to obtain the monic irreducible polynomial

Reviewer: Maybe make it explicit that the input to this algorithm is a seed (I stumbled upon this in the keygen and did at first not make a connection to this section)

:math:`g` from a random element :math:`\beta \in \mathbb{F}_{q} [y]/F(y)`, as
described in Step 3 of Section 8.1 in [CMCE-ISO]_. This arithmetic
is implemented in the ``Classic_McEliece_Polynomial_Ring`` class, representing
the polynomial ring :math:`\mathbb{F}_{q} [y]/F(y)` and supporting the
multiplication of two polynomials.

> Following the recommendation in the implementors' guide Section 6.1 that

Author: Yeah, I'll add a new reference [CMCE-IMPL] and use it in the other places referring to this document.

accompanies [CMCE-R4]_, the minimal polynomial is computed by finding a unique
solution to the equation :math:`g_0\beta^0 + ... + g_{t-1}\beta^{t-1} = \beta^t`.
A constant-time Gauss elimination algorithm is used to solve this equation.
The algorithm aborts if the solution is non-unique. The minimal polynomial
is then represented as a ``Classic_McEliece_Minimal_Polynomial`` object, which
includes logic for serialization and deserialization as described in
Section 9.2.9 of [CMCE-ISO]_.

.. _pubkey/cmce/bitvector:

Bit Vector Arithmetic
^^^^^^^^^^^^^^^^^^^^^

Botan's ``bitvector`` class is a versatile tool for working with bits.
One such application is the Classic McEliece algorithm, where it handles

> binary matrices, error vectors, code words, and the column selections in the private

Reviewer: Do you mean the control bits?

Author: I think column selection is the correct term (i.e. the result of `move_columns`). We don't really use them for control bit generations.

Reviewer: True. I think I meant that we also use `bitvector `for the control bits storage, i.e., we read them as bitvectors and store them as bitvectors. It's minor but maybe we can add "control bit storage" to the sentence.

Author: Fair point, I'll add it 👍

key. The ``bitvector`` class provides a range of useful methods for these tasks.

A ``bitvector`` can be dynamically initialized with any length. It can be created
from a vector of bytes and serialized back to it, as described in Section 9.2.1
of [CMCE-ISO]_. The class allows for adding or removing bits, accessing and
manipulating individual bits, and obtaining specific properties of the vector,
such as the Hamming weight or checking if the vector is all-zero. Sub-vectors
can also be extracted from an existing ``bitvector``.

Furthermore, the ``bitvector`` class supports binary operations between two vectors
of the same length, including AND, OR, and XOR. These operations are optimized
for performance, especially in the context of Classic McEliece.

> The design of the ``bitvector`` class prioritizes constant times, particularly

Author: The constant time operations are mentioned in the text below. I agree that terms like prioritize and 'ct in mind' are too hand-wavy. It's copilot's fault :P

submatrix `T` is stored in the matrix object, mirroring its representation in
the public key specified in Section 9.2.7 of [CMCE-ISO]_.

Classic McEliece instances with the suffix ``f`` employ a semi-systematic transformation
strategy, utilizing the parameters :math:`(\mu, \nu) = (32, 64)`, as described
in Section 7.2.3 of [CMCE-ISO]_. Following the recommendation in Section 6.1 of
the implementation guide accompanying [CMCE-R4]_, a Gauss algorithm is executed
to create an identity matrix for the first :math:`mt-\mu` rows. Subsequently, a
modified Gauss algorithm achieves a reduced row-echelon form for
the :math:`\mu \times \nu` submatrix at position :math:`(\mu, \mu)`.
This process determines the column selection, i.e., the indices of the non-zero
columns. As Section 7.2.3 Step 5 of [CMCE-ISO]_ outlines, the matrix's
columns and the field ordering are permuted according to the pivots. The main
Gauss algorithm concludes at this point, leaving the matrix in its modified
and systematic form.

For encapsulation, the matrix :math:`H` is multiplied by an error vector
:math:`e`. This multiplication is performed by computing the parity of the

> Hamming weight of :math:`e\ \textrm{XOR}\ r` for each row :math:`r` of :math:`H`.

Author: ⊕k!

Matrix creation and multiplication are implemented in constant time for matrices
with fixed dimensions.

.. _pubkey/cmce/keys_internal:

Key Pair
^^^^^^^^

Botan's key pair for Classic McEliece consists of two classes:
``Classic_McEliece_PrivateKeyInternal`` and ``Classic_McEliece_PublicKeyInternal``.
As defined in Section 9.2.12, the private key stores the key generation seed,
column selection, monic irreducible polynomial, field ordering control bits,
and the seed for implicit rejection. On the other hand, the public key
contains the sub-matrix :math:`T` of the binary parity check matrix
:math:`H = (I_{mt}|T)`.

The class ``Classic_McEliece_KeyPair_Internal`` holds both the private and public
keys and contains key generation method. Details are discussed in Section
:ref:`Key Generation <pubkey/cmce/key_gen>`.

.. _pubkey/cmce/encaps_internal:

> Encapsulation Internals

Reviewer: There's this loop going to MAX_ATTEMPTS in raw_kem_encrypt to prevent endless loops. Should we mention this and the corresponding failure probabilities here?

Author: Yeah, good point!

^^^^^^^^^^^^^^^^^^^^^^^

The class ``Classic_McEliece_Encryptor`` implements Botan's key

> encapsulation interface. Performing encapsulation requires two building blocks:

Author: IMO `under a given public key` adds no relevant information and may be a bit confusing. I think I'll leave it as it is...

Fixed weight vector creation and error vector encoding.

The algorithm described in Section 8.4 of [CMCE-ISO]_ is followed to create a
fixed weight error vector. Random elements :math:`d_0,...,d_{\tau-1}` are
generated, where the first :math:`t` elements smaller than :math:`n` are selected as
:math:`a_0,...,a_{t-1}`. It is important to note that selecting

> :math:`d_i` elements for :math:`a_j` elements may leak through timing

Author: I mean it is leaked which d_i is selected for which a_j. I admit my version sounds a bit weird. I'll reword it, thanks!

side-channels. However, the content of these elements remains undisclosed,
preventing attackers from gaining helpful information from this side channel.

Next, all possible pairs are compared to ensure that all elements of
:math:`a_0,...,a_{t-1}` are distinct. Finally, a binary error vector :math:`e`
is created with set bits at the locations of :math:`a_0,...,a_{t-1}`. This

> creation process also considers constant times.

Author: I consider writing _satisfies_ instead to satisfy your consideration.

> creation process also considers constant times.

Reviewer: After some satisfying consideration of your consideration I consider my consideration satisfied.

For encoding, the parity check matrix :math:`H` is multiplied with :math:`e` as
Section :ref:`Matrix Operations <pubkey/cmce/matrix>` describes. The
encapsulation algorithm used in Botan is outlined in Section
:ref:`Key Encapsulation <pubkey/cmce/encapsulation>`.

.. _pubkey/cmce/decaps_internal:

Decapsulation Internals
^^^^^^^^^^^^^^^^^^^^^^^

The class ``Classic_McEliece_Decryptor`` in Botan is responsible for
key decapsulation. One of the crucial steps in the decapsulation algorithm of
Classic McEliece is the decoding subroutine described in Section 7.4 of
[CMCE-ISO]_. This subroutine is implemented based on the recommendations
provided in [McBits]_. The decoding process utilizes Berlekamp's algorithm for
Goppa decoding.

To begin with, the code word :math:`C` that needs to be decoded is extended by

> appending zeros. This results in a binary vector :math:`\nu = (C,0,\dots,0) \in \mathbb{F}_2^{n}`,

Author: Ops, you are right! Good catch!

as Section 7.2 Step 1 of [CMCE-ISO]_ defines. Subsequently, the syndrome for
Berlekamp's method is computed from :math:`\nu`. The syndrome is a vector given by

> :math:`\left( \sum\nolimits_{i} \frac{\nu_i\alpha_i^0}{g(\alpha_i)^2},\dots,\sum\nolimits_{i} \frac{\nu_i\alpha_i^{n-1}}{g(\alpha_i)^2} \right)`,

Author: I think these details are especially relevant because the specification does not provide a specific algorithm for Goppa decoding. Also, the description in the McBits paper is quite hidden.

where :math:`\alpha_i` are the first :math:`n` field ordering elements and
:math:`g` is the Goppa polynomial.

> Next, the error locator polynomial :math:`\sigma` is computed using the

Author: I think you are right. _An error locator_ sounds better.

Berlekamp-Massey algorithm on the syndrome. The resulting polynomial has a
particular property concerning the error vector :math:`e` we search for.
Specifically, :math:`\sigma(\alpha_i) = 0` if and only if
:math:`e_i = 1`. By evaluating :math:`\sigma` at :math:`\alpha_0,\dots,\alpha_{n-1}`,
we can reconstruct the error vector :math:`e`.

To ensure accurate decoding, Botan follows the recommendation in Section 6.3 of
the implementation guide of [CMCE-R4]_. It computes the syndrome for the error
vector :math:`e` and compares it with the syndrome for :math:`\nu`. If the
comparison holds and the weight of :math:`e` is equal to :math:`t`, we consider

> the decoding successful. Otherwise, it is flagged as a failure.

Author: Almost. The difference is, that the syndrome is not the Goppa Code's syndrome, but the syndrome used in Berlekamp's Algorithm. Implementation Guide Section 6.3:
> There are other parity-check matrices $H'$ (related to $\hat{H}$  in key generation) for the same code
that are recovered from $\Gamma'$ much more efficiently than MatGen, and that can be applied to
vectors without using quadratic space; see generally [14]. Comparing $H'(v + e)$ to 0, where
$v = (C, 0, \dots, 0)$, has the same effect as comparing $C$ to $He$.

I guess our syndromes are $H'v$ and $H'e$, while Step 4 of [CMCE-ISO] uses the original syndromes $Hv$ and $He$.

It is worth noting that the syndrome computation, Berlekamp-Massey algorithm, and
locator polynomial evaluation are implemented in constant time. Additionally,
the checks for the weight of :math:`e` and the syndrome comparison are designed

> to avoid early abortion if any check fails.

Author: It also hides if the decoding was sucessful or not. I'll try to make it more clear, thanks.

the basis for deriving all the necessary parameters used within the algorithm.
They are collected in a ``Classic_McElice_Parameters`` object, which includes
all the parameters defined in the specifications [CMCE-ISO]_ and [CMCE-R4]_.
This object is passed to all algorithm components, ensuring consistent
parameter usage.

.. _pubkey/cmce/key_gen:

Key Generation
--------------

Classic McEliece key generation follows Section 8.3 of [CMCE-ISO]_ and is
implemented within the ``Classic_McEliece_KeyPair_Internal`` class. It works as
follows:

.. admonition:: Classic McEliece Key Generation

   **Input:**

   -  ``rng``: random number generator

   **Output:**

   -  ``SK``, ``PK``: private and public key

   **Steps:**

   1. Generate a random value ``seed`` using ``rng``
   2. ``s, ordering_bits, irreducible_bits, next_seed = PRF(seed)``
   3. | Create a field ordering ``field_ordering`` using ``ordering_bits``
      | On failure, set ``seed = next_seed`` and go to step 2

   4. | Create a monic irreducible polynomial ``g`` using ``irreducible_bits``
      | On failure, set ``seed = next_seed`` and go to step 2

>    5. | Create a parity check matrix in systematic form ``H = (I_mt | T)`` using ``field_ordering`` and ``g``. Meanwhile, compute the column selection ``c``

Reviewer: ```suggestion
   5. | Create a parity check matrix in systematic form ``H = (I_mt | T)`` using ``Classic_McEliece_Matrix::create_matrix_and_apply_pivots`` on ``field_ordering`` and ``g``. During this process, the column selection ``c`` is also computed.
```
I think at these crucial points we can use some function names? (We can also keep it like this if the resulting text is too cluttered).

Also, to me, "meanwhile" suggests concurrency.

Author: I'm not sure if the actual method name is useful for any reader. Also, I fear it's indeed too cluttered. We could add it to the internal section if we really want to provide this information. I'll leave it as it is for now.

I'll change the _meanwhile_ term, though. Thanks!

>       | On failure, set ``seed = next_seed`` and go to step 2

Reviewer: Should we mention somewhere how failures are represented in our implementation?

Author: Do you mean by using `std::optional`? IMO, that's too much technically detailed.

   6. ``SK = {seed, c, g, field_ordering, s}, PK = {T}``

   **Notes:**

   - ``PRG`` is an application of ``SHAKE256`` with an input prefix byte 64.
     The output length is ``n/8 + 4q + 2t + 32`` bytes. It is defined
     in Section 9.1 of [CMCE-ISO]_.
   - Only the first ``n`` elements of the field ordering are used to
     create the parity check matrix.
   - For instances with the suffix ``f``, the semi-systematic transformation
     strategy is employed. In this case, ``field_ordering`` is updated to
     reflect the column selection.
   - To store the private and public keys as bytes, the respective entries are
     serialized following the specifications of Section 9.2 of [CMCE-ISO]_.

.. _pubkey/cmce/encapsulation:

Key Encapsulation
-----------------

The Classic McEliece encapsulation procedure of Botan follows Section 8.5 of
[CMCE-ISO]_ and works as follows:

.. admonition:: Classic McEliece Encapsulation

   **Input:**

   - ``PK = {T}``: public key
   - ``rng``: random number generator

   **Output:**

   - ``encap_key``: ciphertext of shared key
   - ``shared_key``: plaintext shared key

   **Steps:**

>    1. Generate a random error vector ``e`` of weight ``t`` using ``rng``

Reviewer: This is fine but I'd be more of a fan to write:

`2. ``H = Classic_McEliece_Matrix`` representing I_mt | T, i.e., to get the previously described names and classes into the pseudocode.

   **Notes:**

   - ``Hash`` is an application of ``SHAKE256`` with 32 output bytes as defined
     in Section 9.1 of [CMCE-ISO]_.

.. _pubkey/cmce/decapsulation:

Key Decapsulation
-----------------

The Classic McEliece decapsulation procedure of Botan follows Section 8.6 of
[CMCE-ISO]_ and works as follows:

.. admonition:: Classic McEliece Decapsulation

   **Input:**

   -  ``SK = {seed, c, g, field_ordering, s}``: secret key
   -  ``encap_key``: encapsulated key bytes

   **Output:**

   -  ``shared_key``: shared key

   **Steps:**

   1. Depending on whether the parameter set includes plaintext confirmation (suffix ``pc``):

       a. **Without pc:** ``c0 = encap_key``
       b. **With pc:** ``c0, c1 = encap_key``, split after ``ceil(m*t/8)`` bytes

   2. | Decode ``c0`` to obtain ``e`` using Berlekamp's algorithm and set ``b = 1``
      | On failure set ``e = s`` and ``b = 0``

   3. | **Only for pc instances:** ``c1_p = Hash(2, e)``
      | If ``c1_p != c1`` set ``e = s`` and ``b = 0``

   4. ``shared_key = Hash(b, e, encap_key)``

   **Notes:**

   - ``Hash`` is an application of ``SHAKE256`` with 32 output bytes as defined
     in Section 9.1 of [CMCE-ISO]_.

>    - The assignments of ``e`` and ``b`` in steps 2 and 3 are implemented using

Reviewer: ```suggestion
   - The failure comparisons and assignments in steps 2 and 3 are implemented using
```

binary matrices, error vectors, code words, and column selections.
The ``bitvector`` class provides a range of useful methods for these tasks.

A ``bitvector`` can be dynamically initialized with any length. It can be created
from a vector of bytes and serialized back to it, as described in Section 9.2.1
of [CMCE-ISO]_. The class allows for adding or removing bits, accessing and
manipulating individual bits, and obtaining specific properties of the vector,
such as the Hamming weight or checking if the vector is all-zero. Sub-vectors
can also be extracted from an existing ``bitvector``.

Furthermore, the ``bitvector`` class supports binary operations between two vectors
of the same length, including AND, OR, and XOR. These operations are optimized
for performance, especially in the context of Classic McEliece.

The design of the ``bitvector`` class supports many side-channel resistant
operations. The time taken for bit accesses and manipulations depends
only on the bit's position, not its value. Operations between two vectors
are also constant-time when the operands have the same length. Other helper
functions used in Classic McEliece, such as Hamming weight computation, are

> also available with side-channel protection.

Reviewer: ```suggestion
also available with side-channel protection.
Botan's Classic McEliece implementation only uses the constant-time functionalities of ``bitvector``.
```

.. _pubkey/cmce:

Classic McEliece
================

Botan implements the Classic McEliece Key Encapsulation Mechanism (KEM) as defined in the
Round 4 NIST submission [CMCE-R4]_ and the ClassicMcEliece ISO Draft [CMCE-ISO]_. It
supports all parameter sets defined in both documents, listed
in Table :ref:`Supported Classic McEliece parameter sets <pubkey/cmce/parameter_table>`.

.. _pubkey/cmce/parameter_table:

> .. table:: Supported Classic McEliece parameter sets (see [CMCE-ISO]_ Section 10 and [CMCE-R4]_ Section 7). ``<f>`` and ``<pc>`` indicate that sets with and without f and pc are supported.

Reviewer 2: ```suggestion
.. table:: Supported Classic McEliece parameter sets (see [CMCE-ISO]_ Section 10 and [CMCE-R4]_ Section 7). ``<f>`` and ``<pc>`` indicate that sets of the "fast" variant (f) and with or without plaintext confirmation (pc) are supported.
```

   | ``mceliece6960119<pc><f>`` | 13        | 6960      | 119       | :math:`z^{13}+z^4+z^3+z+1` | :math:`y^{119}+y^8+1`           |
   +----------------------------+-----------+-----------+-----------+----------------------------+---------------------------------+
   | ``mceliece8192128<pc><f>`` | 13        | 8192      | 128       | :math:`z^{13}+z^4+z^3+z+1` | :math:`y^{128}+y^7+y^2+y+1`     |
   +----------------------------+-----------+-----------+-----------+----------------------------+---------------------------------+

.. _pubkey/cmce/key_generation:

Algorithm Internals
-------------------

The implementation of Classic McEliece is organized based on the algorithm's
operations and structures. Table :ref:`Classic McEliece components
<pubkey/cmce/component_table>` shows an overview of the components and
their corresponding file locations. Classic McEliece, like all Key Encapsulation
Mechanisms (KEMs), consists of three main algorithms: Key generation, key
encapsulation, and key decapsulation. In the case of Classic McEliece, the key
generation process involves creating a field ordering and an irreducible

> polynomial. These two components define the randomly chosen Goppa code, the

Reviewer 2: ```suggestion
polynomial. These two components define a randomly chosen Goppa code which is the
```

secret information stored in the private key. A parity check matrix in
systematic form is derived from the Goppa code to generate the public key used
for encapsulation. More detailed information about these processes can be found
in the following sections.

.. _pubkey/cmce/component_table:

> .. table:: Classic McEliece components and file locations. Its files are located at :srcref:`src/lib/pubkey/classic_mceliece/`.

Reviewer 2: ```suggestion
.. table:: Classic McEliece components and file locations. The files are located at :srcref:`src/lib/pubkey/classic_mceliece/`.
```

   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+
   | :ref:`Galois Field Arithmetic <pubkey/cmce/gf>`        | :srcref:`[src/lib/pubkey/classic_mceliece]/cmce_gf.h`            | Galois field data structure and logic                                      |
   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+
   | :ref:`Field Ordering <pubkey/cmce/field_ordering>`     | :srcref:`[src/lib/pubkey/classic_mceliece]/cmce_field_ordering.h`| Field ordering creation and Beneš network computation                      |
   +--------------------------------------------------------+------------------------------------------------------------------+----------------------------------------------------------------------------+

>    | :ref:`Polynomial Arithmetic <pubkey/cmce/poly>`        | :srcref:`[src/lib/pubkey/classic_mceliece]/cmce_poly.h`          | Polynomial data structure, logic, and minimal polynomial creation          |

Reviewer 3: Die Links zeigen hier scheinbar auf den Rohde-Schwarz fork von Botan und ich bekomme dort die Fehlermeldung: "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository."

locator polynomial evaluation are implemented in constant time. Additionally,
the checks for the weight of :math:`e` and the syndrome comparison are designed
to avoid early abortion if any check fails. This ensures that no information about
the decoding success leaks.

.. _pubkey/cmce/params:

Parameters
^^^^^^^^^^

The ``Classic_McEliece_Parameter_Set`` enum contains all instances of the
Classic McEliece algorithm listed in Table :ref:`Supported Classic McEliece
Parameter Sets <pubkey/cmce/parameter_table>`. These parameter sets serve as
the basis for deriving all the necessary parameters used within the algorithm.
They are collected in a ``Classic_McElice_Parameters`` object, which includes
all the parameters defined in the specifications [CMCE-ISO]_ and [CMCE-R4]_.
This object is passed to all algorithm components, ensuring consistent
parameter usage.

.. _pubkey/cmce/key_gen:

Key Generation
--------------

Classic McEliece key generation follows Section 8.3 of [CMCE-ISO]_ and is
implemented within the ``Classic_McEliece_KeyPair_Internal`` class. It works as
follows:

.. admonition:: Classic McEliece Key Generation

>    **Input:**

Reviewer 2: From our side, it would be useful to list parameters or mode specifications as an input in the documentation when they are also required as an input to the respective implemented function like it is for Classic McEliece.
